File: cloud-run-bot/scrapers/trendyol_scraper.js
/**
 * Trendyol Scraper (Node.js port)
 * Enhanced for full compatibility & robust extraction
 */
const BaseProductScraper = require('./base_scraper');

class TrendyolScraper extends BaseProductScraper {
  get domain() { return 'trendyol.com'; }

  canHandle(url) {
    const lower = url.toLowerCase();
    return lower.includes('trendyol.com') || lower.includes('ty.gl');
  }

  scrapeImage($, url) {
    // 1. JSON-LD şemasından (Öncelikli)
    const product = this.findProductJsonLd($);
    if (product && product['image']) {
      const img = this.extractImageFromProductJson(product['image']);
      if (img && !this.isLogoUrl(img)) {
        const resolved = this.resolveImageUrl(img, url);
        if (resolved) return resolved;
      }
    }

    // 2. Open Graph / Meta Tags
    const metaImages = [
      $('meta[property="og:image"]').attr('content'),
      $('meta[name="twitter:image"]').attr('content')
    ];
    for (const ogImg of metaImages) {
      if (ogImg && !this.isLogoUrl(ogImg)) {
        const resolved = this.resolveImageUrl(ogImg, url);
        if (resolved) return resolved;
      }
    }

    // 3. DOM Seçicileri
    const imgSelectors = [
      '.product-image-container img',
      '.detail-main-img img',
      'img.main-img',
      '.gallery-container img',
      '.base-product-image img',
      '[data-testid="product-image"] img'
    ];
    for (const sel of imgSelectors) {
      const imgElements = $(sel);
      for (let i = 0; i < imgElements.length; i++) {
        const el = $(imgElements[i]);
        const src = el.attr('src') || el.attr('data-src') || el.attr('data-original');
        if (src && !this.isLogoUrl(src)) {
          const resolved = this.resolveImageUrl(src, url);
          if (resolved) return resolved;
        }
      }
    }
    return null;
  }

  scrapeTitle($) {
    // 1. JSON-LD şemasından (Öncelikli)
    const product = this.findProductJsonLd($);
    if (product && product['name']) {
      const name = product['name'].toString().trim();
      if (name.length > 0) return name;
    }

    // 2. DOM Seçicileri
    const titleSelectors = [
      '[data-testid="product-title"]',
      'h1.pr-new-br',
      '.product-title',
      'h1.product-title',
      'h1'
    ];
    for (const sel of titleSelectors) {
      const el = $(sel).first();
      if (el.length) {
        const txt = el.text().trim();
        if (txt.length > 0) return txt;
      }
    }

    // 3. Meta Tags
    const ogTitle = $('meta[property="og:title"]').attr('content');
    if (ogTitle && ogTitle.trim()) return ogTitle.trim();

    return null;
  }

  scrapePriceLabel($) {
    // 1. DOM Seçicileri: Trendyol Plus fiyat/başlık öğeleri
    const plusSelectors = [
      '.ty-plus-price-header',
      '.ty-plus-price',
      '[class*="ty-plus-price"]',
      '[class*="ty-plus"]',
      '[class*="plus-price"]',
      '.ty-plus-banner-desktop'
    ];

    for (const sel of plusSelectors) {
      const el = $(sel).first();
      if (el.length > 0) {
        const text = el.text().trim();
        if (/plus/i.test(text) || sel.includes('ty-plus')) {
          return "Plus'a Özel";
        }
      }
    }

    // 2. Metin bazlı DOM araması (Plus'a Özel)
    let foundText = false;
    $('span, div, p, b, strong, a, label, h1, h2, h3').each((_, el) => {
      const text = $(el).clone().children().remove().end().text().trim();
      if (/(?:trendyol\s*)?plus['’]?\s*a\s*özel/i.test(text)) {
        foundText = true;
        return false;
      }
    });
    if (foundText) return "Plus'a Özel";

    // 3. Script etiketleri & initial state kontrolü
    let scriptFound = false;
    const scriptPlusRegex = /ty-plus|hasPlusPromotion|isPlusExclusive|plusPromotion|(?:trendyol\s*)?plus(?:\\u0027|['’])\s*a\s*özel/i;
    $('script').each((_, el) => {
      const text = $(el).text();
      if (text && scriptPlusRegex.test(text)) {
        scriptFound = true;
        return false;
      }
    });
    if (scriptFound) return "Plus'a Özel";

    return null;
  }

  scrapePrice($) {
    // 1. Plus indirimli fiyat DOM kontrolü (En spesifik güncel fiyat)
    const plusPriceEl = $('.ty-plus-price-discounted-price, .ty-plus-price .ty-plus-price-discounted-price, [class*="ty-plus-price-discounted-price"]').first();
    if (plusPriceEl.length) {
      const val = this.parsePriceText(plusPriceEl.text());
      if (val !== null && val > 0) return val;
    }

    // 2. JSON-LD şemasından (Öncelikli)
    const productJson = this.findProductJsonLd($);
    if (productJson) {
      const priceLd = this.extractPriceFromProductJson(productJson);
      if (priceLd && priceLd > 0) {
        return priceLd;
      }
    }

    // 3. DOM Seçicileri (Fallback 1)
    const priceSelectors = [
      '.discounted',
      '.prc-dsc',
      '.price-container span',
      '.prc-slg',
      '.pr-bx-w .prc-dsc',
      '.product-price-container span',
      '.prc-box-dsc',
      '.prc-box-sll',
      '[class*="price-discounted"]',
      '[class*="prc-dsc"]',
      '[class*="selling-price"]'
    ];
    for (const sel of priceSelectors) {
      const el = $(sel).first();
      if (el.length) {
        const val = this.parsePriceText(el.text());
        if (val !== null && val > 0) return val;
      }
    }

    // 4. Script Search (Fallback 2: Initial State / Next Data)
    let foundPrice = null;
    $('script').each((_, el) => {
      const text = $(el).text();
      if (text && (text.includes('__PRODUCT_DETAIL_APP_INITIAL_STATE__') || text.includes('product":{') || text.includes('__NEXT_DATA__'))) {
        try {
          const m = text.match(/"(?:discountedPrice|sellingPrice|price|salePrice)"\s*:\s*\{[^\}]*?"value"\s*:\s*([\d.]+)/);
          if (m) {
            const val = parseFloat(m[1]);
            if (!isNaN(val) && val > 0) {
              foundPrice = val;
              return false;
            }
          }
        } catch (_) {}
      }
    });

    return foundPrice;
  }

  scrapeOriginalPrice($, currentPrice) {
    if (!currentPrice || currentPrice <= 0) return null;

    let candidates = [];

    // 1. DOM selectors for old / strikethrough / original prices (Öncelikli)
    const domSelectors = [
      '.old-price',
      '.prc-org',
      '.ty-plus-price-original-price',
      '[class*="price-original"]',
      '[class*="original-price"]',
      '[class*="prc-org"]',
      '[class*="old-price"]',
      '.prc-box-org',
      'del',
      's'
    ];

    for (const selector of domSelectors) {
      $(selector).each((_, el) => {
        const txt = $(el).text().trim();
        const parsed = this.parsePriceText(txt);
        if (parsed !== null && parsed > currentPrice) {
          candidates.push(parsed);
        }
      });
    }

    if (candidates.length > 0) {
      candidates = candidates.filter(c => c > currentPrice && c <= currentPrice * 5);
      if (candidates.length > 0) {
        candidates.sort((a, b) => a - b);
        return candidates[0];
      }
    }

    // 2. JSON-LD High Price (Fallback 1)
    const productJson = this.findProductJsonLd($);
    if (productJson && productJson['offers']) {
      const offers = Array.isArray(productJson['offers']) ? productJson['offers'] : [productJson['offers']];
      for (const offer of offers) {
        if (offer['highPrice']) {
          const hp = parseFloat(offer['highPrice']);
          if (!isNaN(hp) && hp > currentPrice) candidates.push(hp);
        }
      }
    }

    // 3. Initial State Script Search (Fallback 2)
    $('script').each((_, el) => {
      const text = $(el).text();
      if (text && (text.includes('__PRODUCT_DETAIL_APP_INITIAL_STATE__') || text.includes('product":{') || text.includes('__NEXT_DATA__'))) {
        try {
          const matches = text.match(/"(?:originalPrice|marketPrice|crossedOutPrices)"\s*:\s*\{[^\}]*?"value"\s*:\s*([\d.]+)/g);
          if (matches) {
            for (const m of matches) {
              const valMatch = m.match(/"value"\s*:\s*([\d.]+)/);
              if (valMatch) {
                const val = parseFloat(valMatch[1]);
                if (!isNaN(val) && val > currentPrice) {
                  candidates.push(val);
                }
              }
            }
          }
        } catch (_) {}
      }
    });

    if (candidates.length === 0) return null;

    candidates = candidates.filter(c => c > currentPrice && c <= currentPrice * 5);
    if (candidates.length === 0) return null;

    candidates.sort((a, b) => a - b);
    return candidates[0];
  }

  scrapeDescription($) {
    const descEl = $('meta[name="description"], meta[property="og:description"]').first();
    return descEl.length ? descEl.attr('content')?.trim() : null;
  }

  scrapeBreadcrumbs($) {
    const title = this.scrapeTitle($) || '';
    const breadcrumbs = this.extractBreadcrumbsFromJsonLd($, title, 'trendyol');
    if (breadcrumbs && breadcrumbs.length > 0) return breadcrumbs;

    // DOM Fallback
    const list = [];
    $('.product-detail-breadcrumb a, .breadcrumb a, .breadcrumbs a').each((_, el) => {
      const text = $(el).text().trim();
      if (text) {
        const lower = text.toLowerCase();
        if (lower !== 'anasayfa' && lower !== 'trendyol' && text !== title && text.length < 50) {
          list.push(text);
        }
      }
    });
    return list;
  }

  scrapeRating($) {
    const product = this.findProductJsonLd($);
    if (product) {
      const rating = this.extractRatingFromProductJson(product);
      if (rating && (rating.ratingValue != null || rating.ratingCount != null)) {
        return rating;
      }
    }

    let ratingValue = null;
    let ratingCount = null;

    $('script').each((_, el) => {
      const text = $(el).text();
      if (text && (text.includes('ratingValue') || text.includes('averageRating'))) {
        if (!ratingValue) {
          const vm = text.match(/"(?:ratingValue|averageRating)"\s*:\s*"?([\d.,]+)"?/);
          if (vm) {
            const parsed = parseFloat(vm[1].replace(',', '.'));
            if (!isNaN(parsed) && parsed > 0 && parsed <= 5) ratingValue = parsed;
          }
        }
        if (!ratingCount) {
          const cm = text.match(/"(?:ratingCount|totalRatingCount|reviewCount)"\s*:\s*"?(\d+)"?/);
          if (cm) {
            const parsed = parseInt(cm[1]);
            if (!isNaN(parsed) && parsed > 0) ratingCount = parsed;
          }
        }
      }
    });

    return { ratingValue, ratingCount };
  }

  scrapeBrand($) {
    // 1. JSON-LD (Öncelikli)
    const product = this.findProductJsonLd($);
    if (product) {
      const brand = this.extractBrandFromProductJson(product);
      if (brand && brand.trim()) return brand.trim();
    }

    // 2. DOM Seçicileri
    const brandSelectors = [
      'h1.pr-new-br a',
      'a.product-brand',
      'span.pr-new-br',
      '.brand-name',
      '[data-testid="brand-name"]'
    ];
    for (const sel of brandSelectors) {
      const el = $(sel).first();
      if (el.length) {
        const txt = el.text().trim();
        if (txt.length > 0) return txt;
      }
    }

    return null;
  }
}

module.exports = TrendyolScraper;

File: lib/services/scrapers/trendyol_scraper.dart
import 'dart:async';
import 'dart:convert';
import 'package:html/dom.dart' as dom;
import 'base_scraper.dart';

class TrendyolScraper extends BaseProductScraper {
  @override
  String get domain => 'trendyol.com';

  @override
  bool canHandle(String url) {
    final lowerUrl = url.toLowerCase();
    return lowerUrl.contains('trendyol.com') || lowerUrl.contains('ty.gl');
  }

  @override
  FutureOr<String?> scrapePriceLabel(dom.Document document) {
    // 1. DOM Seçicileri: Trendyol Plus gerçek fiyat & banner elemanları
    const plusSelectors = [
      '.ty-plus-price-header',
      '.ty-plus-price-discounted-price',
      '.ty-plus-price',
      '.ty-plus-banner-desktop',
    ];

    for (final sel in plusSelectors) {
      final el = document.querySelector(sel);
      if (el != null) {
        final text = el.text.trim();
        if (text.isNotEmpty) {
          return "Plus'a Özel";
        }
      }
    }

    // 2. Metin bazlı DOM araması (Plus'a Özel / Trendyol Plus'a Özel)
    final plusRegex = RegExp(r"^(?:trendyol\s*)?plus['’]?\s*(?:a\s*özel|la\s*daha\s*az\s*öde|ile)", caseSensitive: false);
    final elements = document.querySelectorAll('span, div, p, b, strong, a, label, h1, h2, h3');
    for (final el in elements) {
      final text = el.text.trim();
      if (text.isNotEmpty && text.length <= 60 && plusRegex.hasMatch(text)) {
        return "Plus'a Özel";
      }
    }

    // 3. Script etiketleri: SADECE açık boolean TRUE koşulları (window condition=true veya JSON true)
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final text = script.text;
      if (text.isNotEmpty) {
        if (RegExp(r'''__envoy_ty-plus-banner__CONDITION["']?\s*\]?\s*=\s*true''', caseSensitive: false).hasMatch(text)) {
          return "Plus'a Özel";
        }
        if (RegExp(r'''"hasPlusPromotion"\s*:\s*true|"isPlusExclusive"\s*:\s*true''', caseSensitive: false).hasMatch(text)) {
          return "Plus'a Özel";
        }
      }
    }

    return null;
  }

  @override
  String? scrape({
    required dom.Document document,
    required String url,
    required bool Function(String urlString) isLogoUrl,
    required String? Function(String? imageUrl, String pageUrl) resolveImageUrl,
    required void Function(String message) log,
  }) {
    // 1. JSON-LD şemasından görsel çekmeyi dene (Öncelikli)
    final productJson = findProductJsonLd(document);
    if (productJson != null && productJson['image'] != null) {
      final imgLd = extractImageFromProductJson(productJson['image']);
      if (imgLd != null && imgLd.isNotEmpty) {
        final resolved = resolveImageUrl(imgLd, url);
        if (resolved != null && !isLogoUrl(resolved)) {
          log('✅ Trendyol görseli JSON-LD ile bulundu: $resolved');
          return resolved;
        }
      }
    }

    // 2. Open Graph (Fallback 1)
    final ogImage = document.querySelector('meta[property="og:image"]')?.attributes['content'];
    if (ogImage != null && ogImage.isNotEmpty) {
      final resolved = resolveImageUrl(ogImage, url);
      if (resolved != null && !isLogoUrl(resolved)) {
        log('✅ Trendyol görseli og:image ile bulundu: $resolved');
        return resolved;
      }
    }

    // 3. Main Product Image (Fallback 2)
    final imgElements = document.querySelectorAll('.product-image-container img, .detail-main-img img, img.main-img');
    for (final img in imgElements) {
      final src = img.attributes['src'] ?? img.attributes['data-src'];
      if (src != null && src.isNotEmpty) {
        final resolved = resolveImageUrl(src, url);
        if (resolved != null && !isLogoUrl(resolved)) {
          log('✅ Trendyol görseli img etiketiyle bulundu: $resolved');
          return resolved;
        }
      }
    }

    return null;
  }

  @override
  String? scrapeTitle(dom.Document document) {
    // 1. JSON-LD şemasından (Öncelikli)
    final productJson = findProductJsonLd(document);
    if (productJson != null && productJson['name'] != null) {
      return productJson['name'].toString().trim();
    }

    // 2. DOM Seçicileri (Fallback)
    final titleEl = document.querySelector('[data-testid="product-title"]') ??
                    document.querySelector('.product-title') ??
                    document.querySelector('h1.product-title');
    if (titleEl != null) {
      return titleEl.text.trim();
    }
    return null;
  }

  @override
  Future<double?> scrapePrice(dom.Document document) async {
    // 1. Plus indirimli fiyat DOM kontrolü (En spesifik güncel fiyat)
    final plusPriceEl = document.querySelector('.ty-plus-price-discounted-price') ??
                        document.querySelector('.ty-plus-price .ty-plus-price-discounted-price') ??
                        document.querySelector('[class*="ty-plus-price-discounted-price"]');
    if (plusPriceEl != null) {
      final val = parsePriceText(plusPriceEl.text);
      if (val != null && val > 0) return val;
    }

    // 2. JSON-LD şemasından (Öncelikli)
    final productJson = findProductJsonLd(document);
    if (productJson != null) {
      final priceLd = extractPriceFromProductJson(productJson);
      if (priceLd != null && priceLd > 0) {
        return priceLd;
      }
    }

    // 3. DOM Seçicileri (Fallback)
    final priceEl = document.querySelector('.discounted') ??
                    document.querySelector('.prc-dsc') ??
                    document.querySelector('.price-container span');
    if (priceEl != null) {
      final val = parsePriceText(priceEl.text);
      if (val != null && val > 0) {
        return val;
      }
    }

    return null;
  }

  @override
  double? scrapeOriginalPrice(dom.Document document, double? currentPrice) {
    if (currentPrice == null || currentPrice <= 0) return null;

    final candidates = <double>[];

    // 1. DOM Seçicileri (Öncelikli)
    final domSelectors = [
      '.old-price',
      '.prc-org',
      '.ty-plus-price-original-price',
      '[class*="price-original"]',
      '[class*="original-price"]',
      '[class*="prc-org"]',
      '[class*="old-price"]',
      'del',
      's',
    ];

    for (final selector in domSelectors) {
      for (final el in document.querySelectorAll(selector)) {
        final parsed = parsePriceText(el.text);
        if (parsed != null && parsed > currentPrice) {
          candidates.add(parsed);
        }
      }
    }

    if (candidates.isNotEmpty) {
      final valid = candidates.where((c) => c > currentPrice && c <= currentPrice * 5).toList();
      if (valid.isNotEmpty) {
        valid.sort();
        return valid.first;
      }
    }

    // 2. Initial State Script Search (Fallback)
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final text = script.text;
      if (text.contains('__PRODUCT_DETAIL_APP_INITIAL_STATE__') || text.contains('product":{')) {
        try {
          final matches = RegExp(r'"(?:originalPrice|sellingPrice|marketPrice)"\s*:\s*\{[^\}]*?"value"\s*:\s*([\d.]+)').allMatches(text);
          for (final m in matches) {
            final raw = m.group(1);
            if (raw != null) {
              final val = double.tryParse(raw);
              if (val != null && val > currentPrice) {
                candidates.add(val);
              }
            }
          }
        } catch (_) {}
      }
    }

    if (candidates.isEmpty) return null;

    final valid = candidates.where((c) => c > currentPrice && c <= currentPrice * 5).toList();
    if (valid.isEmpty) return null;

    valid.sort();
    return valid.first;
  }

  @override
  double? scrapeRatingValue(dom.Document document) {
    print('[aggregateRating] TrendyolScraper: ratingValue aranıyor...');
    final productJson = findProductJsonLd(document);
    if (productJson != null) {
      final rating = extractRatingFromProductJson(productJson);
      if (rating?['ratingValue'] != null) {
        final val = (rating!['ratingValue'] as num).toDouble();
        print('[aggregateRating] TrendyolScraper: JSON-LD ile ratingValue bulundu: $val');
        return val;
      }
    }

    // Fallback: Raw Script Regex Arama
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final text = script.text;
      if (text.contains('aggregateRating') || text.contains('ratingValue')) {
        final match = RegExp(r'"ratingValue"\s*:\s*"?([\d.,]+)"?').firstMatch(text);
        if (match != null) {
          final raw = match.group(1)?.replaceAll(',', '.');
          final parsed = raw != null ? double.tryParse(raw) : null;
          if (parsed != null) {
            print('[aggregateRating] TrendyolScraper: Regex fallback ile ratingValue bulundu: $parsed');
            return parsed;
          }
        }
      }
    }

    print('[aggregateRating] TrendyolScraper: ratingValue bulunamadı (null)');
    return null;
  }

  @override
  int? scrapeRatingCount(dom.Document document) {
    print('[aggregateRating] TrendyolScraper: ratingCount/reviewCount aranıyor...');
    final productJson = findProductJsonLd(document);
    if (productJson != null) {
      final rating = extractRatingFromProductJson(productJson);
      if (rating?['ratingCount'] != null) {
        final cnt = (rating!['ratingCount'] as num).toInt();
        print('[aggregateRating] TrendyolScraper: JSON-LD ile ratingCount/reviewCount bulundu: $cnt');
        return cnt;
      }
    }

    // Fallback: Raw Script Regex Arama
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final text = script.text;
      if (text.contains('aggregateRating') || text.contains('reviewCount') || text.contains('ratingCount')) {
        final match = RegExp(r'"(?:ratingCount|reviewCount)"\s*:\s*"?(\d+)"?').firstMatch(text);
        if (match != null) {
          final parsed = int.tryParse(match.group(1) ?? '');
          if (parsed != null) {
            print('[aggregateRating] TrendyolScraper: Regex fallback ile ratingCount/reviewCount bulundu: $parsed');
            return parsed;
          }
        }
      }
    }

    print('[aggregateRating] TrendyolScraper: ratingCount bulunamadı (null)');
    return null;
  }

  @override
  String? scrapeBrand(dom.Document document) {
    print('[aggregateRating] TrendyolScraper: brand (marka) aranıyor...');
    final productJson = findProductJsonLd(document);
    if (productJson != null) {
      final brand = extractBrandFromProductJson(productJson);
      if (brand != null && brand.isNotEmpty) {
        print('[aggregateRating] TrendyolScraper: JSON-LD ile brand bulundu: $brand');
        return brand;
      }
    }

    // Fallback: Raw Script Regex Arama
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final text = script.text;
      if (text.contains('"brand"') || text.contains('"Brand"')) {
        final match = RegExp(r'"brand"\s*:\s*\{\s*"@type"\s*:\s*"Brand"\s*,\s*"name"\s*:\s*"([^"]+)"').firstMatch(text) ??
                      RegExp(r'"brand"\s*:\s*"([^"]+)"').firstMatch(text);
        if (match != null) {
          final bName = match.group(1)?.trim();
          if (bName != null && bName.isNotEmpty && !bName.contains('{')) {
            print('[aggregateRating] TrendyolScraper: Regex fallback ile brand bulundu: $bName');
            return bName;
          }
        }
      }
    }

    print('[aggregateRating] TrendyolScraper: brand bulunamadı (null)');
    return null;
  }

  @override
  List<String> scrapeBreadcrumbs(dom.Document document) {
    final scripts = document.querySelectorAll('script');
    final productTitle = scrapeTitle(document) ?? '';

    for (final script in scripts) {
      final type = script.attributes['type']?.trim().toLowerCase();
      if (type == 'application/ld+json') {
        try {
          final sanitizedText = script.text.replaceAll('\r\n', ' ').replaceAll('\n', ' ').replaceAll('\r', ' ');
          final data = jsonDecode(sanitizedText);
          final breadcrumbs = _extractBreadcrumbsFromJson(data, productTitle);
          if (breadcrumbs.isNotEmpty) {
            return breadcrumbs;
          }
        } catch (_) {}
      }
    }

    // DOM Fallback
    final breadcrumbElements = document.querySelectorAll('.product-detail-breadcrumb a, .breadcrumb a, .breadcrumbs a');
    if (breadcrumbElements.isNotEmpty) {
      final List<String> list = [];
      for (final el in breadcrumbElements) {
        final text = el.text.trim();
        if (text.isNotEmpty) {
          final lower = text.toLowerCase();
          if (lower != 'anasayfa' && lower != 'trendyol' && text != productTitle && text.length < 50) {
            list.add(text);
          }
        }
      }
      if (list.isNotEmpty) return list;
    }

    return [];
  }

  List<String> _extractBreadcrumbsFromJson(dynamic json, String productTitle) {
    if (json is Map) {
      // 1. BreadcrumbList kontrolü
      if (json['@type'] == 'BreadcrumbList' || json['@type'] == 'http://schema.org/BreadcrumbList') {
        final items = json['itemListElement'];
        if (items is List) {
          final List<String> breadcrumbs = [];
          for (final item in items) {
            if (item is Map) {
              String? name;
              if (item['name'] != null) {
                name = item['name'].toString().trim();
              } else if (item['item'] is Map && item['item']['name'] != null) {
                name = item['item']['name'].toString().trim();
              }

              if (name != null && name.isNotEmpty) {
                final lowerName = name.toLowerCase();
                if (lowerName != 'anasayfa' && lowerName != 'trendyol') {
                  // Ürün başlığının kendisini veya çok uzun ürün isimlerini breadcrumb olarak ekleme
                  if (name != productTitle && name.length < 50) {
                    breadcrumbs.add(name);
                  }
                }
              }
            }
          }
          if (breadcrumbs.isNotEmpty) return breadcrumbs;
        }
      }

      // 2. Product category kontrolü
      if (json['@type'] == 'Product' || json['@type'] == 'http://schema.org/Product' ||
          json['@type'] == 'ProductGroup' || json['@type'] == 'http://schema.org/ProductGroup') {
        final categoryField = json['category'];
        if (categoryField is String && categoryField.isNotEmpty) {
          final parts = categoryField
              .split(RegExp(r'\s*>\s*'))
              .map((e) => e.trim())
              .where((e) => e.isNotEmpty)
              .where((e) {
                final lower = e.toLowerCase();
                return lower != 'anasayfa' && lower != 'trendyol' && e != productTitle && e.length < 50;
              })
              .toList();
          if (parts.isNotEmpty) return parts;
        }
      }

      // Recursive arama
      for (final value in json.values) {
        if (value is Map || value is List) {
          final res = _extractBreadcrumbsFromJson(value, productTitle);
          if (res.isNotEmpty) return res;
        }
      }
    } else if (json is List) {
      for (final item in json) {
        final res = _extractBreadcrumbsFromJson(item, productTitle);
        if (res.isNotEmpty) return res;
      }
    }
    return [];
  }
}
